### PR TITLE
Allow plan sharing for uncorrelated subqueries

### PR DESCRIPTION
**Pull request by dey4ss**
Imagine the following SQL query (simple join between the TPC-DS tables `web_sales` and `date_dim` as seen in many queries):
```sql
SELECT *
  FROM web_sales, date_dim
 WHERE ws_ship_date_sk = d_date_sk

```
Given that every value of `ws_ship_date_sk` is also in `d_date_sk`, we could (and soon will) rewrite this query into something like the following:
```SQL
SELECT *
  FROM web_sales
 WHERE ws_ship_date_sk BETWEEN (SELECT MIN(d_date_sk) FROM date_dim)
                           AND (SELECT MAX(d_date_sk) FROM date_dim)
```

The query plan for this query could looke like this (with efficient operator reusage):

```
          +-------------------+
          |   PredicateNode   |
          | A.a BETWEEN       |
          |   <subquery1> AND |
          |   <subquery2>     | * * * * * * * * * * * * * 
          +-------------------+                           *
                  /            *                           *
                 /              *                           *
                /                * uncorrelated              * uncorrelated
               /                 * subquery1                 * subquery2
              /                  *                           *
             /         +-------------------+       +-------------------+
            /          |  ProjectionNode   |       |  ProjectionNode   |
           /           |      MIN(b)       |       |      MAX(b)       |
          |            +-------------------+       +-------------------+
          |                             \              /
          |                          +-------------------+
          |                          |   AggregateNode   |
          |                          |  MIN(b), MAX(b)   |
          |                          +-------------------+
          |                                    |
+-------------------+                +-------------------+
| StoredTableNode A |                | StoredTableNode B |
+-------------------+                +-------------------+
```

For this LQP, the optimizer (in debug mode) complains that `AggregateNode`'s outputs are part of different LQPs since each subquery is considered separately in `Optimizer::validate_lqp(...)`. This PR changes this behavior and treats uncorrelated subqueries just like normal inputs. For correlated subqueries, the desired validation error is still thrown (the result of correlated subqueries depends on the parameter value and cannot be reused).

### CODE REVIEW

**Comment by Bouncner** [comment]
What is the issue with correlated subqueries? If "the base" is equal, I don't see why we should not allow plan sharing here. Do you a counter example where it crashes?

**Comment by dey4ss** [comment]
> What is the issue with correlated subqueries? If "the base" is equal, I don't see why we should not allow plan sharing here. Do you a counter example where it crashes?

Still thinking of an example, but I think one can create one where we the subquery LQP is the same, the correlated parameter column is the same, but the actual *value* of the parameter is different (i.e., there are different filters on the parameter). We cannot check this during optimization since the parameter value is only available in the execution. Anyway, we require correlated subqueries to be completely managed by the `ExpressionEvaluator` that *must* make a deep copy of the subquery PQP for each row. Thus, sharing sublans (and reusing the results) does not have an effect at all.

File: src/test/lib/logical_query_plan/lqp_utils_test.cpp [hunk 1]
```diff
@@ -298,38 +301,62 @@ TEST_F(LQPUtilsTest, LQPInsertAboveNode) {
...
+  // Prepare an LQP with multiple subqueries in a nested manner.
 
-  // clang-format off
-  const auto nested_subquery_lqp =
-  AggregateNode::make(expression_vector(), expression_vector(max_(a_a)),
-    node_a);
-  const auto max_a_subquery = lqp_subquery_(nested_subquery_lqp);
+  for (const auto only_correlated_subqueries : {true, false}) {
+    auto message = std::stringstream{};
+    message << "Consider only uncorrelated subqueries: " << std::boolalpha << only_correlated_subqueries;
```
Comment (Bouncner) [inline src/test/lib/logical_query_plan/lqp_utils_test.cpp:308]:
```
SCOPED_TRACE(std::string{"Consider only uncorrelated subqueries: "} + (only... ? "true" : "false"));
```
?

Comment (Bouncner) [reply src/test/lib/logical_query_plan/lqp_utils_test.cpp:308]: Not that it would matter ...

File: src/lib/optimizer/optimizer.cpp [hunk 2]
```diff
@@ -149,118 +285,30 @@ std::shared_ptr<AbstractLQPNode> Optimizer::optimize(
...
-  // If you can think of a way in which an LQP can be corrupt, please add it!
-
-  // First, collect all LQPs (the main LQP and all subqueries)
+  // First, collect all LQPs (the main LQP and all uncorrelated subqueries).
   auto lqps = std::vector<std::shared_ptr<AbstractLQPNode>>{root_node};
-  auto subquery_expressions_by_lqp = collect_lqp_subquery_expressions_by_lqp(root_node);
-  for (const auto& [lqp, subquery_expressions] : subquery_expressions_by_lqp) {
+  auto subquery_expressions_by_lqp = collect_lqp_subquery_expressions_by_lqp(root_node, true);
+  for (const auto& [lqp, _] : subquery_expressions_by_lqp) {
     lqps.emplace_back(lqp);
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/lib/optimizer/optimizer.cpp:299]: Micro opt inside `if constexpr (HYRISE_DEBUG) {...}`? Okay, I do as you please.

File: src/lib/optimizer/optimizer.hpp [hunk 3]
```diff
@@ -1,16 +1,13 @@
 #pragma once
 
-#include <algorithm>
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/lib/optimizer/optimizer.hpp:3]: Well, I'm thankful C++ lets us include headers multiple times so we do not have to create weird include collections to ensure everything is included from bottom to top. However, I do not see much worth in including every header in every file if it is already included somewhere in another included header. So whenever I see unnecessary includes, I remove them.

File: src/test/lib/logical_query_plan/lqp_utils_test.cpp (hunk 1 above)
Comment (dey4ss) [reply src/test/lib/logical_query_plan/lqp_utils_test.cpp:308]: actually a debug artifact. Keep it as `SCOPED_TRACE` or remove?

File: src/test/lib/logical_query_plan/lqp_utils_test.cpp [hunk 4]
```diff
@@ -298,38 +301,62 @@ TEST_F(LQPUtilsTest, LQPInsertAboveNode) {
...
-    PredicateNode::make(greater_than_(b_x, max_a_subquery),
-      node_b));
-  const auto x_greater_than_max_a_subquery = lqp_subquery_(subquery_lqp);
+    // clang-format off
+    const auto nested_subquery_lqp =
+    AggregateNode::make(expression_vector(), expression_vector(max_(a_a)),
+      node_a);
+    const auto max_a_subquery = lqp_subquery_(nested_subquery_lqp);
+
+
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/logical_query_plan/lqp_utils_test.cpp:317]: weird that this was not resolved by `format.sh`.

File: src/test/lib/optimizer/optimizer_test.cpp [hunk 5]
```diff
@@ -84,14 +104,39 @@ TEST_F(OptimizerTest, AssertsInPlanReferences) {
   // clang-format off
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/optimizer/optimizer_test.cpp:104]: The test checks that we throw an error if an expression used by a node (`x` in that case) comes from the same LQP. Tested code is here:
https://github.com/hyrise/hyrise/blob/5adcbd2bbae08c1150eb277fd99bfe3cc788370e/src/lib/optimizer/optimizer.cpp#L94-L110

Since the checks drill down if the original node referenced by the LQPColumnExpression is in the same LQP, it's kind of what it says. Maybe `AssertsOriginalNodeInSamePlan`?

Now to the second question: It worked before. However, the used uncorrelated subquery contains `node_b`, which outputs `x`. Since we treat uncorrelated subqueries as regular nodes, they are now also part of the mapping LQP root -> all nodes in the LQP. Thus, it would not throw an error anymore.

File: src/lib/optimizer/optimizer.cpp [hunk 6]
```diff
@@ -149,118 +285,30 @@ std::shared_ptr<AbstractLQPNode> Optimizer::optimize(
...
   }
 
-  std::unordered_map<std::shared_ptr<const AbstractLQPNode>, std::unordered_set<std::shared_ptr<const AbstractLQPNode>>>
-      nodes_by_lqp;
+  auto nodes_by_lqp = LQPNodesByLQP{};
+  // Second, assign each LQPNode to its LQP. Uncorrelated subqueries are treated like normal LQPNodes since their output
+  // can also be used multiple times.
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/lib/optimizer/optimizer.cpp:304]: Correlated subquery LQPs must not be reused, they are completely managed by the ExpressionEvaluator. Thus, they cannot have multiple "outputs", i.e., must not be part of multiple LQPSubqueryExpressions. For a long time, we were a bit shy about subqueries and treated uncorrelated subqueries the same, which was unnecessary. Since the results of uncorrelated subquery LQPs do not depend on the value of a correlated parameter, we can use them in a thousand subquery expressions if we like, which is exactly what is the case for all "regular" LQP nodes, which can have an unlimited number of outputs. Recently, we started to treat uncorrelated subqueries more like regular LQPNodes/Operators, we schedule them like any other task, etc. So no need to treat them differently here.

tl;dr: `Since the results of uncorrelated subquery LQPs do not depend on correlated parameters, we can use them in multiple LQPSubqueryExpressions and treat them like normal LQPNodes. However, correlated subqueries must not share nodes with any other (sub-)LQP.`?

**Comment by dey4ss** [comment]
I love how you age your reviews like fine wine <3 (weeks :b)

File: src/test/lib/optimizer/optimizer_test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/test/lib/optimizer/optimizer_test.cpp:104]: `AssertsOriginalNodeInSamePlan` sounds better to me.

File: src/lib/optimizer/optimizer.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/optimizer/optimizer.cpp]: I like the comment.

Should we also name that the expr eval owns the correlated LQPs and deep-copies them?

**Review by Bouncner — APPROVED** [review]
